### PR TITLE
Only set healthcheck user agents when the ping-user-agent is set, and don't check blank user agents against healthcheck user agents

### LINKED ISSUE #649: No User-agent is considered a healthcheck User-agent when none specified
## Expected Behavior
When I make a request without a user agent, I expect this to be treated like a normal request. This is the case with oauth2-proxy 5.1.1.

e.g.: on 5.1.1
```
GET /my-url HTTP/1.0

HTTP/1.0 403 Forbidden
Date: Thu, 02 Jul 2020 05:28:20 GMT
Content-Type: text/html; charset=utf-8
```

## Current Behavior
When no User-agent header is set, and no ping user agent options are specified (the default option), oauth2-proxy returns the "OK" healthcheck ping response.

e.g.:
```
GET /my-url HTTP/1.0

HTTP/1.0 200 OK
Date: Thu, 02 Jul 2020 05:33:16 GMT
Content-Length: 2
Content-Type: text/plain; charset=utf-8

OKConnection closed by foreign host.
```

When the `ping-user-agent` option is specified, e.g.: to value "blah" this does not occur.

## Possible Solution
A workaround is to specify the `ping-user-agent` option.

The full solution requires explaining how this happens. A blank user agent is considered == to an empty string. When no `-ping-user-agent` option is specified, this is considered to be an empty string.

This reveals two problems:
* When no `ping-user-agent` is specified, `main.go` sets up a health check user agent of ""
* When no user agent is specified, the empty string is still checked against the health check user agents.

I have a pull request incoming to solve this issue.

## Steps to Reproduce (for bugs)

1. Run a simple oauth2-proxy server with `go build && ./oauth2-proxy --cookie-secret="1234567891234567" --client-id="mysecret" --client-secret="mysecret" --email-domain="mydomain.com"`
2. Connect via telnet, with `telnet localhost 4180`
3. Enter `GET /my-url HTTP/1.0` and press enter twice.
4. You'll see an OK response rather than an oauth2-proxy login page

## Context
GitLab web-hooks don't specify a User-agent, meaning GitLab web-hooks couldn't get to our upstream application.

## Your Environment
macOS and Ubuntu Linux

- Version used: 6.0.

### PR DESCRIPTION
**Pull request by jordancrawfordnz**
## Description
A blank user agent is considered == to an empty string. When no -ping-user-agent option is specified, this is considered to be an empty string.

This reveals two problems:
- When no ping-user-agent is specified, main.go sets up a health check user agent of ""
- When no user agent is specified, the empty string is still checked against the health check user agents.

Now, the health check user agents are only set if ping-user-agent is specified.
In addition, empty user agents are not checked against healthcheck user agents.

Additional tests have been added to verify these situations.

## Motivation and Context
Resolves #649.

## How Has This Been Tested?
Newly introduced automated testing and manually verifying it against the reproduction steps in #649.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by jordancrawfordnz** [comment]
Hi reviewers - I'm open to suggestions about how to resolve the Codeclimate error - I'm thinking I'd need to extract some of the logic into seperate functions within `main.go` but that may not be common place to do that?

I'm normally a Ruby dev so this Go world is quite new to me.

**Comment by jordancrawfordnz** [comment]
Thanks for the review - when I get a chance (hopefully later today for me) I'll address the feedback.

**Comment by jordancrawfordnz** [comment]
> I am hesitant about checking the length of the user agent before doing the map look up, 99% of requests will have a user agent set, there are very few cases where it will be empty.

Yep, that is fair, requests need to be as quick as possible. We can opt to just remove blank use agents on initialisation.

> Secondly, GitLab really really should be setting a user agent, there is an open issue for it but it doesn't seem to be doing anything at the moment. Normally if there's no user agent, it's because the user is trying to hide themselves, in which case, they're almost certainly a bad actor!

Yes, absolutely! Hopefully it gets fixed in GitLab.

> Also, I believe we will have similar issues for the path, would you add a couple of test cases for the request path being empty as well? (Though I'm pretty sure the request path should never be empty for a valid request)

Yes that's a good point - I attempted to test this myself with telnet but of course it didn't take the shape of a valid HTTP request.

File: main.go [hunk 1]
```diff
@@ -83,7 +83,10 @@ func main() {
 	}
 
 	healthCheckPaths := []string{opts.PingPath}
-	healthCheckUserAgents := []string{opts.PingUserAgent}
+	healthCheckUserAgents := []string{}
+	if len(opts.PingUserAgent) > 0 {
```
[the remark this answers is not in the document]
Comment (jordancrawfordnz) [reply main.go:87]: This is fair - avoids the error for code complexity too.

File: pkg/middleware/healthcheck.go [hunk 2]
```diff
@@ -41,8 +41,10 @@ func isHealthCheckRequest(paths, userAgents map[string]struct{}, req *http.Reque
 	if _, ok := paths[req.URL.EscapedPath()]; ok {
 		return true
 	}
-	if _, ok := userAgents[req.Header.Get("User-Agent")]; ok {
-		return true
+	if len(req.Header.Get("User-Agent")) > 0 {
```
[the remark this answers is not in the document]
Comment (jordancrawfordnz) [reply pkg/middleware/healthcheck.go:44]: This no longer makes sense if the healthcheck middleware is responsible for removing the blank entries. The reason for this is that I didn't 'trust' `main.go` prevent construction of the middleware without blank entries since this had no tests enforcing it.

File: pkg/middleware/healthcheck_test.go [hunk 3]
```diff
@@ -50,6 +50,22 @@ var _ = Describe("HealthCheck suite", func() {
...
+			headers:               map[string]string{},
+			expectedStatus:        404,
+			expectedBody:          "404 page not found\n",
+		}),
+		Entry("with no health check user agents configured", &requestTableInput{
+			healthCheckPaths:      []string{"/ping"},
+			healthCheckUserAgents: []string{},
+			requestString:         "http://example.com/different",
+			headers:               map[string]string{},
+			expectedStatus:        404,
```
[the remark this answers is not in the document]
Comment (jordancrawfordnz) [reply pkg/middleware/healthcheck_test.go:66]: Have added this now - I've treated this (and one for path too) as a general happy path check that it can handle not being configured.

**Comment by jordancrawfordnz** [comment]
@JoelSpeed - ready for re-review when you get a chance
